File: javascript/easy/imageSmoother.js
/**
 * @param {number[][]} M
 * @return {number[][]}
 */
var imageSmoother = function(M) {
    const dirs = [[0,-1],[0,1],[-1,0],[1,0],[-1,-1],[1,1],[-1,1],[1,-1]];
    const res = [];
    for(let i = 0; i < M.length; i++) {
        const temp = [];
        for(let j = 0; j < M[i].length; j++) {
            temp.push(getAverage(M, i, j, dirs));
        }
        res.push(temp);
    }
    return res;
};

var getAverage = function(M, i, j, dirs) {
    let sum = M[i][j], count = 1;
    for(let dir of dirs) {
        let x = (i + dir[0]), y = (j + dir[1]);
        if(isValid(M, x, y)) {
            sum += M[x][y];
            count++;
        }
    }
    return parseInt(sum/count);
};

var isValid = function(M, i, j) {
    return i >= 0 && i < M.length && j >= 0 && j < M[i].length;    
};

File: ruby/easy/image_smoother.rb
# @param {Integer[][]} m
# @return {Integer[][]}
def image_smoother(m)
    dirs = [[0,-1],[0,1],[-1,0],[1,0],[-1,-1],[1,1],[-1,1],[1,-1]]
    res = Array.new()
    for i in 0..m.length-1
        temp = Array.new()
        for j in 0..m[i].length-1
            temp[j] = avg(m, i, j, dirs)
        end
        res.push(temp)
    end
    return res
end

def avg(m, i, j, dirs)
    sum = m[i][j]
    count = 1
    dirs.each do |dir|
        x = i + dir[0]
        y = j + dir[1]
        if valid(m, x, y)
            sum += m[x][y]
            count += 1
        end
    end
    return (sum/count)
end

def valid(m, x, y)
    return x >= 0 && x < m.length && y >= 0 && y < m[x].length
end
